Stop retrieving the packageName from the manifest

When upgrading to AGP 7.2.0, one of the "recommended steps" of the upgrade assistant is to "Move package from Android manifest to build files" :
> Declaration of a project's namespace using the package attribute of the Android manifest is deprecated in favour of a namespace declaration in build files.

If we proceed as recommended, it breaks the plugin with the following stacktrace
```
Caused by: java.lang.NullPointerException
	at fr.sjcqs.task.SettingsGenerationTaskKt.getPackageNameFromManifest(SettingsGenerationTask.kt:113)
	at fr.sjcqs.task.SettingsGenerationTaskKt.access$getPackageNameFromManifest(SettingsGenerationTask.kt:1)
	at fr.sjcqs.task.ConfigurationGenerationTask$Companion$create$1.execute(SettingsGenerationTask.kt:86)
	at fr.sjcqs.task.ConfigurationGenerationTask$Companion$create$1.execute(SettingsGenerationTask.kt:75)
	at org.gradle.api.internal.tasks.DefaultTaskContainer.create(DefaultTaskContainer.java:362)
	at fr.sjcqs.task.ConfigurationGenerationTask$Companion.create(SettingsGenerationTask.kt:84)
	at fr.sjcqs.konfig.KonfigPlugin$createTasks$1.execute(KonfigPlugin.kt:18)
	at fr.sjcqs.konfig.KonfigPlugin$createTasks$1.execute(KonfigPlugin.kt:10)
```

It fails because the packageName used by the plugin is extracted from the manifest. Instead, the plugin should use [the namespace attribute](https://developer.android.com/studio/build/configure-app-module#set-namespace), the same attribute used to generate `R.java` and `BuildConfig.java`

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 [Unreleased]: https://github.com/sjcqs/gradle-konfig/compare/1.0.0...HEAD
+### Added
+- Support the namespace attribute from `android` extension to set the Configuration package name.
 
 ## [1.0.7] - 2022-05-13
 [1.0.7]: https://github.com/sjcqs/gradle-konfig/releases/tag/1.0.7

--- a/gradle.properties
+++ b/gradle.properties
@@ -1,5 +1,5 @@
 #Fri May 13 09:40:10 CEST 2022
-androidBuildToolsVersion=7.1.0
+androidBuildToolsVersion=7.2.0
 snakeYamlVersion=1.30
 pluginId=fr.sjcqs.konfig
 pluginGroup=fr.sjcqs.konfig
@@ -8,7 +8,7 @@ jupiterVersion=5.8.2
 org.gradle.jvmargs=-Xmx4G
 testRepositoryDir=test-repository
 kotlinVersion=1.6.21
-pluginVersion=1.0.7
+pluginVersion=1.0.8
 kotlin.code.style=official
 pluginName=konfig
 testRepositoryName=test

--- a/konfig/src/main/java/fr/sjcqs/konfig/KonfigPlugin.kt
+++ b/konfig/src/main/java/fr/sjcqs/konfig/KonfigPlugin.kt
@@ -2,23 +2,38 @@ package fr.sjcqs.konfig
 
 import fr.sjcqs.task.ConfigurationGenerationTask
 import fr.sjcqs.task.TaskConfiguration
-import fr.sjcqs.task.TaskConfigurationProvider
+import fr.sjcqs.task.TaskConfigurationFactory
 import fr.sjcqs.utils.get
+import fr.sjcqs.utils.pathJoin
 import org.gradle.api.Plugin
 import org.gradle.api.Project
+import java.io.File
 
 open class KonfigPlugin : Plugin<Project> {
     override fun apply(target: Project) {
-        val configuration = TaskConfigurationProvider.getConfiguration(target)
+        val configuration = TaskConfigurationFactory.getConfiguration(target)
         target.createTasks(configuration)
     }
 
     private fun Project.createTasks(configuration: TaskConfiguration) {
         configuration.variants.all { variant ->
-            val task = ConfigurationGenerationTask.create(project, variant)
-            val outputDirectory = task.outputDir.get().asFile
+            val outputDirectory = project.file(getOutputDirectoryName(project.buildDir, variant.dirName))
+            val task = ConfigurationGenerationTask.register(
+                project,
+                variant,
+                configuration.namespaceProvider,
+                outputDirectory
+            )
             variant.registerJavaGeneratingTask(task, outputDirectory)
             configuration.sourceSets[variant.name].java.srcDir(outputDirectory)
         }
+    }
+
+    private fun getOutputDirectoryName(buildDir: File, variantDirectoryName: String): String {
+        return pathJoin(buildDir.absolutePath, GENERATED_DIRECTORY, variantDirectoryName)
+    }
+
+    companion object {
+        private const val GENERATED_DIRECTORY = "generated/source/konfig"
     }
 }

--- a/konfig/src/main/java/fr/sjcqs/task/ConfigurationGenerationTask.kt
+++ b/konfig/src/main/java/fr/sjcqs/task/ConfigurationGenerationTask.kt
@@ -16,6 +16,7 @@ import org.gradle.api.InvalidUserDataException
 import org.gradle.api.Project
 import org.gradle.api.file.ConfigurableFileCollection
 import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.provider.Provider
 import org.gradle.api.tasks.CacheableTask
 import org.gradle.api.tasks.Input
 import org.gradle.api.tasks.InputFiles
@@ -23,6 +24,7 @@ import org.gradle.api.tasks.OutputDirectory
 import org.gradle.api.tasks.PathSensitive
 import org.gradle.api.tasks.PathSensitivity
 import org.gradle.api.tasks.TaskAction
+import org.gradle.api.tasks.TaskProvider
 import org.gradle.work.Incremental
 import org.gradle.work.InputChanges
 import org.yaml.snakeyaml.Yaml
@@ -70,17 +72,25 @@ abstract class ConfigurationGenerationTask @Inject constructor() : DefaultTask()
 
     companion object {
         private const val DEFAULT_CONFIG_NAME = "default"
-        private const val GENERATED_DIRECTORY = "generated/source/konfig"
         private const val INPUT_DIRECTORY = "config"
         private val LOAD_CLASS = mutableMapOf<String, Any?>()::class
 
         @Throws(InvalidUserDataException::class)
-        fun create(project: Project, variant: BaseVariant): ConfigurationGenerationTask {
+        fun register(
+            project: Project,
+            variant: BaseVariant,
+            namespaceProvider: Provider<String?>,
+            targetDirectory: File
+        ): TaskProvider<ConfigurationGenerationTask> {
             val name = "generate${variant.name}${Token.Root.ROOT_KEY}"
-            return project.tasks.create(name, ConfigurationGenerationTask::class.java) { task ->
+            return project.tasks.register(name, ConfigurationGenerationTask::class.java) { task ->
                 task.apply {
-                    packageName = variant.getPackageNameFromManifest()
-                    outputDir.set(project.file(getOutputDirectoryName(project.buildDir, variant.dirName)))
+                    /* We use the module namespace then we fall back to the packageName declared in the manifest*/
+                    packageName = namespaceProvider.get() ?: variant.getPackageNameFromManifest() ?: error(
+                        "Failed to found the module packageName.\n" +
+                            "Either declare a namespace in build.gradle or a packageName in your manifest file."
+                    )
+                    outputDir.set(targetDirectory)
 
                     val flavorName = variant.flavorName
                     val buildType = variant.buildType.name
@@ -107,15 +117,12 @@ abstract class ConfigurationGenerationTask @Inject constructor() : DefaultTask()
             }
         }
 
-        private fun getOutputDirectoryName(buildDir: File, variantDirectoryName: String): String {
-            return pathJoin(buildDir.absolutePath, GENERATED_DIRECTORY, variantDirectoryName)
+        private fun BaseVariant.getPackageNameFromManifest(): String? {
+            val manifestFile = sourceSets.map { it.manifestFile }
+                .firstOrNull(File::exists)
+                ?: return null
+            val node = XmlParser().parse(manifestFile)
+            return node.attribute("package")?.toString().orEmpty()
         }
     }
-}
-
-private fun BaseVariant.getPackageNameFromManifest(): String {
-    val manifestFile = sourceSets.map { it.manifestFile }
-        .first(File::exists)
-    val node = XmlParser().parse(manifestFile)
-    return node.attribute("package").toString()
 }

--- a/konfig/src/main/java/fr/sjcqs/task/TaskConfiguration.kt
+++ b/konfig/src/main/java/fr/sjcqs/task/TaskConfiguration.kt
@@ -4,8 +4,10 @@ import com.android.build.gradle.api.AndroidSourceSet
 import com.android.build.gradle.api.BaseVariant
 import org.gradle.api.DomainObjectCollection
 import org.gradle.api.NamedDomainObjectContainer
+import org.gradle.api.provider.Provider
 
 interface TaskConfiguration {
     val variants: DomainObjectCollection<out BaseVariant>
+    val namespaceProvider: Provider<String?> // lazily provide namespace
     val sourceSets: NamedDomainObjectContainer<AndroidSourceSet>
 }

--- a/konfig/src/main/java/fr/sjcqs/task/TaskConfigurationFactory.kt
+++ b/konfig/src/main/java/fr/sjcqs/task/TaskConfigurationFactory.kt
@@ -7,8 +7,9 @@ import com.android.build.gradle.api.BaseVariant
 import org.gradle.api.DomainObjectCollection
 import org.gradle.api.NamedDomainObjectContainer
 import org.gradle.api.Project
+import org.gradle.api.provider.Provider
 
-object TaskConfigurationProvider {
+object TaskConfigurationFactory {
     private const val APPLICATION_PLUGIN_ID: String = "com.android.application"
     private const val LIBRARY_PLUGIN_ID: String = "com.android.library"
 
@@ -25,6 +26,7 @@ object TaskConfigurationProvider {
         private val extension = project.extensions.getByType(AppExtension::class.java)
         override val variants: DomainObjectCollection<out BaseVariant>
             get() = extension.applicationVariants
+        override val namespaceProvider: Provider<String?> = project.provider { extension.namespace }
         override val sourceSets: NamedDomainObjectContainer<AndroidSourceSet>
             get() = extension.sourceSets
     }
@@ -33,6 +35,7 @@ object TaskConfigurationProvider {
         private val extension = project.extensions.getByType(LibraryExtension::class.java)
         override val variants: DomainObjectCollection<out BaseVariant>
             get() = extension.libraryVariants
+        override val namespaceProvider: Provider<String?> = project.provider { extension.namespace }
         override val sourceSets: NamedDomainObjectContainer<AndroidSourceSet>
             get() = extension.sourceSets
     }

--- a/konfig/src/main/java/fr/sjcqs/utils/MapMerger.kt
+++ b/konfig/src/main/java/fr/sjcqs/utils/MapMerger.kt
@@ -27,7 +27,7 @@ class MapMerger(private val logger: Logger) {
                     logger.d("$key: merging two maps $value and $otherValue")
                     map[key] = deepMerge((value as KonfigMap), otherValue as KonfigMap)
                 } else if (value is List<*> && otherValue is List<*>) {
-                    logger.d("$key: merging two list ${value.joinToString()} with ${otherValue.joinToString()}")
+                    logger.d("$key: merging two list $value with $otherValue")
                     map[key] = value.merge(otherValue)
                 } else {
                     logger.d("$key: overriding $value with $otherValue")

--- a/test-app/app/build.gradle
+++ b/test-app/app/build.gradle
@@ -46,8 +46,8 @@ android {
 
 dependencies {
     implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlinVersion"
-    implementation 'androidx.appcompat:appcompat:1.1.0'
-    implementation 'androidx.constraintlayout:constraintlayout:1.1.3'
+    implementation 'androidx.appcompat:appcompat:1.4.1'
+    implementation 'androidx.constraintlayout:constraintlayout:2.1.4'
     implementation project(':lib')
-    testImplementation 'junit:junit:4.12'
+    testImplementation 'junit:junit:4.13.2'
 }

--- a/test-app/app/src/main/AndroidManifest.xml
+++ b/test-app/app/src/main/AndroidManifest.xml
@@ -1,6 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
-<manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    package="com.polidea.configsample">
+<manifest xmlns:android="http://schemas.android.com/apk/res/android">
 
     <application
         android:allowBackup="true"

--- a/test-app/build.gradle
+++ b/test-app/build.gradle
@@ -8,10 +8,7 @@ buildscript {
             name = "test"
             url = uri("../test-repository")
         }
-        maven {
-            url 'https://plugins.gradle.org/m2/'
-            content { includeGroupByRegex("org.jetbrains.*") }
-        }
+        gradlePluginPortal()
         google()
         mavenCentral()
     }

